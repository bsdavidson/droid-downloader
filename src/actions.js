export const DOWNLOAD_FILE = "DOWNLOAD_FILE";
export const DOWNLOAD_FOLDER = "DOWNLOAD_FOLDER";
export const REFRESH_DEVICES = "REFRESH_DEVICES";
export const REFRESH_DEVICE_FILES = "REFRESH_DEVICE_FILES";
export const SET_DEVICE = "SET_DEVICE";
export const SET_DEVICES = "SET_DEVICES";
export const SET_DEVICE_FILE = "SET_DEVICE_FILE";
export const SET_DEVICE_FILES = "SET_DEVICE_FILES";
export const SET_DEVICE_PATH = "SET_DEVICE_PATH";
export const SET_FILE_PREVIEW_IMAGE = "SET_FILE_PREVIEW_IMAGE";
export const SET_LOCAL_PATH = "SET_LOCAL_PATH";
export const TOGGLE_FILE_PREVIEW = "TOGGLE_FILE_PREVIEW";
export const TRAVERSE_DEVICE_PATH = "TRAVERSE_DEVICE_PATH";

export function downloadFile(name) {
  return {type: DOWNLOAD_FILE, name};
}

export function downloadFolder() {
  return {type: DOWNLOAD_FOLDER};
}

export function refreshDeviceFiles() {
  return {type: REFRESH_DEVICE_FILES};
}

export function refreshDevices() {
  return {type: REFRESH_DEVICES};
}

export function setDevice(device) {
  return {type: SET_DEVICE, device};
}

export function setDeviceFile(deviceFile) {
  return {type: SET_DEVICE_FILE, deviceFile};
}

export function setDeviceFiles(deviceFiles) {
  return {type: SET_DEVICE_FILES, deviceFiles};
}

export function setDevicePath(devicePath) {
  return {type: SET_DEVICE_PATH, devicePath};
}

export function setDevices(devices) {
  return {type: SET_DEVICES, devices};
}

export function setfilePreviewImage(filePreviewImage) {
  return {type: SET_FILE_PREVIEW_IMAGE, filePreviewImage};
}

export function setLocalPath(localPath) {
  return {type: SET_LOCAL_PATH, localPath};
}

export function toggleFilePreviewFullscreen() {
  return {type: TOGGLE_FILE_PREVIEW};
}

export function traverseDevicePath(dirName) {
  return {type: TRAVERSE_DEVICE_PATH, dirName};
}
