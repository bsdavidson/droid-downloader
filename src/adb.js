import {spawn} from "child_process";
import fs from "fs-extra";
import path from "path";

class ADBExitError extends Error {
  constructor(code, signal, stderr) {
    super(`ADB exited with code ${code} (signal: ${signal}) - ${stderr}`);
    this.code = code;
    this.signal = signal;
    this.stderr = stderr;
  }
}

class Device {
  constructor(serial, state, properties) {
    this.serial = serial;
    this.state = state;
    this.properties = properties;
  }

  static parseLine(line) {
    const fields = line.split(/ +/);
    return new Device(
      fields[0],
      fields[1],
      fields.slice(2).reduce((p, s) => {
        const [k, v] = s.split(":");
        p[k] = v; // eslint-disable-line no-param-reassign
        return p;
      }, {})
    );
  }
}

class DirLine {
  constructor(line, type, timestamp, name) {
    this.type = type;
    this.timestamp = timestamp;
    this.name = name;
    this.line = line;
  }

  static parseLine(line) {
    const parts = line
      .trim()
      .match(
        /^([^ ]+) +(?:[^ ]+ +){2,4}(\d{4}-\d{2}-\d{2} +\d{2}:\d{2}) +(.+)$/
      );
    if (!parts) {
      return null;
    }

    return new DirLine(
      line,
      parts[1].charAt(0) === "d" || parts[1].charAt(0) === "l"
        ? "directory"
        : "file",
      parts[2],
      parts[3].trim()
    );
  }
}

function shellQuote(value) {
  // (╯°□°）╯︵ ┻━┻
  return "'" + value.replace("'", `'"'"'`) + "'"; // eslint-disable-line prefer-template, quotes
}

export function getDevices() {
  return new Promise((resolve, reject) => {
    const process = spawn("/usr/local/bin/adb", ["devices", "-l"]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        reject(new ADBExitError(code, signal, stderrBuffer));
        return;
      }
      resolve(
        stdoutBuffer
          .trim()
          .split("\n")
          .slice(1)
          .map(Device.parseLine)
      );
    });
  });
}

export function getDirectoryList(deviceSerial, dirPath) {
  return new Promise(resolve => {
    const process = spawn("/usr/local/bin/adb", [
      "-s",
      deviceSerial,
      "shell",
      "ls",
      "-l",
      shellQuote(dirPath)
    ]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        // It is expected that ls will exit with an error code when it
        // encounters files or directories that it doesn't have permission
        // to access.
        console.warn(new ADBExitError(code, signal, stderrBuffer));
      }
      resolve(
        stdoutBuffer
          .trim()
          .split("\n")
          .map(DirLine.parseLine)
          .filter(d => d)
      );
    });
  });
}

export function getFileList(rootPath) {
  return new Promise((resolve, reject) => {
    let fileList = [];
    fs.readdir(rootPath, (err, names) => {
      if (err) {
        reject(err);
        return;
      }
      let remaining = names.length;
      function done(delta) {
        remaining -= delta;
        if (remaining === 0) {
          fileList.sort();
          resolve(fileList);
        }
      }
      names.forEach(name => {
        const filePath = path.join(rootPath, name);
        fs.lstat(filePath, (statErr, stats) => {
          if (statErr) {
            console.log(`error in lstat for ${filePath}: ${statErr}`);
          } else if (stats.isDirectory()) {
            getFileList(filePath).then(items => {
              fileList = fileList.concat(items);
              done(1);
            });
            return; // prevent done() from being called too soon
          } else if (stats.isFile()) {
            fileList.push(filePath);
          }
          done(1);
        });
      });
      done(0);
    });
  });
}

export function pull(deviceSerial, source, destination) {
  console.log(deviceSerial, source, destination);
  return new Promise((resolve, reject) => {
    const process = spawn("/usr/local/bin/adb", [
      "-s",
      deviceSerial,
      "pull",
      source,
      destination
    ]);
    let stdoutBuffer = "";
    let stderrBuffer = "";

    process.stdout.on("data", data => {
      stdoutBuffer += data.toString();
    });

    process.stderr.on("data", data => {
      stderrBuffer += data.toString();
    });

    process.on("close", (code, signal) => {
      if (code) {
        reject(new ADBExitError(code, signal, stderrBuffer));
        return;
      }
      resolve(stdoutBuffer.trim().split("\n"));
    });
  });
}
