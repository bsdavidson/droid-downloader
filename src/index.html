<!DOCTYPE html>
<html lang="en">

<head>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no">
  <link href="./fontawesome/css/all.css" type="text/css" rel="stylesheet">
  <link href="./styles/index.scss" type="text/scss" rel="stylesheet">
</head>

<body>
  <div id="App"></div>

  <script>
    require('./index')
  </script>
</body>

</html>