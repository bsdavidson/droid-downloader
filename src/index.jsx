import React from "react";
import ReactDOM from "react-dom";
import {createStore, applyMiddleware} from "redux";
import thunk from "redux-thunk";
import {Provider} from "react-redux";
import {AppContainer} from "react-hot-loader";

import {reducer} from "./reducers";

const store = createStore(reducer, applyMiddleware(thunk));

const render = () => {
  const App = require("./app").default;
  ReactDOM.render(
    <Provider store={store}>
      <AppContainer>
        <App />
      </AppContainer>
    </Provider>,
    document.getElementById("App")
  );
};

render();
if (module.hot) {
  module.hot.accept(render);
}
