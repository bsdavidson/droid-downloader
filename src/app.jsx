import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {ipcRenderer} from "electron";

import {refreshDevices, setLocalPath} from "./actions";

import DeviceToolbar from "./device_toolbar";
import DeviceList from "./device_list";
import DirectoryList from "./directory_list";
import FileInfo from "./file_info";

export class App extends React.Component {
  static handleLocalPathSetClick() {
    ipcRenderer.send("open-file-dialog");
  }

  componentDidMount() {
    this.props.refreshDevices();

    ipcRenderer.on("selected-directory", (event, selectedPath) => {
      this.props.onLocalPathChange(selectedPath);
      // organizeFiles(selectedPath, selectedPath).catch(erroredFiles => {
      //   console.log("erroredFiles", erroredFiles);
      // });
    });
  }

  render() {
    return (
      <div className="app">
        <div className="app-device-list">
          <DeviceList />
        </div>
        <div className="app-directory-list">
          <DeviceToolbar />
          <DirectoryList />
          <FileInfo />
        </div>
      </div>
    );
  }
}

App.propTypes = {
  refreshDevices: PropTypes.func.isRequired,
  onLocalPathChange: PropTypes.func.isRequired
};

function mapDispatchToProps(dispatch) {
  return {
    refreshDevices: () => dispatch(refreshDevices()),
    onLocalPathChange: localPath => dispatch(setLocalPath(localPath))
  };
}

export default connect(
  null,
  mapDispatchToProps
)(App);
