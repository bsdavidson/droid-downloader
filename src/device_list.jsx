import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {setDevice, refreshDevices} from "./actions";
import DroidPropTypes from "./prop_types";

export class DeviceList extends React.Component {
  render() {
    const {devices, device} = this.props;

    return (
      <div className="device-list">
        {devices.length > 0 ? (
          devices.map(d => (
            <div
              className={`device-list-item device-list-item-${
                d.serial === device ? "active" : "inactive"
              }`}
              key={d.serial}>
              <a
                href="#device"
                className="device-list-item-link"
                onClick={() => this.props.onDeviceClick(d.serial)}>
                <i className="device-list-item-icon fas fa-mobile-alt" />
                <span className="device-list-item-model">
                  {d.properties.model.replace(/_/g, " ")}
                </span>
              </a>
            </div>
          ))
        ) : (
          <span className="device-list-nodevices">No Devices Found</span>
        )}
        <div className="device-list-refresh">
          <a
            className="device-list-refresh-link"
            href="#refresh"
            onClick={this.props.onRefreshDevicesClick}>
            <i className="device-list-refresh-icon fas fa-sync-alt" />
            Refresh
          </a>
        </div>
      </div>
    );
  }
}

DeviceList.propTypes = {
  device: DroidPropTypes.device.isRequired,
  devices: DroidPropTypes.devices.isRequired,
  onDeviceClick: PropTypes.func.isRequired,
  onRefreshDevicesClick: PropTypes.func.isRequired
};

function mapStateToProps(state) {
  return {
    devices: state.devices,
    device: state.device
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onDeviceClick: device => dispatch(setDevice(device)),
    onRefreshDevicesClick: () => dispatch(refreshDevices())
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DeviceList);
