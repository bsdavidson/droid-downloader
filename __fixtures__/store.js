import {Device, DirLine} from "../src/adb";

export const device = "ffffbeefface";

export const deviceFiles = [
  new DirLine(
    "directory",
    "2018-01-02 09:45",
    "folder_1",
    "drwxr-xr-x",
    "3001"
  ),
  new DirLine("file", "2017-09-20 21:22", "file_1", "-rw-r--r--", "42100"),
  new DirLine("file", "2017-10-17 01:45", "file_2", "-rw-r--r--", "3125995")
];

export const deviceFile = deviceFiles[1];

export const devicePath = ["storage", "emulated", "0", "DCIM"];

export const devices = [
  new Device("ffff1337face", "device", {
    model: "SM_G955U"
  }),
  new Device("ffffbeefface", "device", {
    model: "SM_G999V"
  }),
  new Device("ffffbadbeef", "device", {
    model: "SM_G910V"
  })
];

export const localPath = "/tmp";
