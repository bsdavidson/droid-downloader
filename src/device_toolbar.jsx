import React, {Component} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {downloadFolder, setDevicePath} from "./actions";
import DroidPropTypes from "./prop_types";
import {App} from "./app";

// const {app} = require("electron").remote;

export class DeviceToolbar extends Component {
  static handleDownloadPathClick() {
    App.handleLocalPathSetClick();
  }

  constructor(props) {
    super(props);

    this.handleBreadCrumbClick = this.handleBreadCrumbClick.bind(this);
  }

  handleBreadCrumbClick(event) {
    this.props.onBreadCrumbPathClick(
      event.target.getAttribute("data-path").split("|#")
    );
  }

  render() {
    if (!this.props.device || this.props.devices.length === 0) {
      return null;
    }

    const device = this.props.devices.filter(
      d => d.serial === this.props.device
    )[0];

    const renderedPath = [
      <span className="directory-list-toolbar-crumb" key="rootdir">
        <span className="directory-list-toolbar-crumb-sep">/</span>
        <a
          href="#dir"
          className="directory-list-toolbar-crumb-link"
          data-path={""}
          onClick={this.handleBreadCrumbClick}>
          root
        </a>
      </span>
    ];

    const crumbPath = [];
    this.props.devicePath.forEach(dir => {
      if (!dir) {
        return;
      }
      crumbPath.push(dir);
      renderedPath.push(
        <span className="directory-list-toolbar-crumb" key={dir}>
          <span className="directory-list-toolbar-crumb-sep">/</span>
          <a
            href="#dir"
            className="directory-list-toolbar-crumb-link"
            data-path={crumbPath.join("|#")}
            onClick={this.handleBreadCrumbClick}>
            {dir}
          </a>
        </span>
      );
    });

    return (
      <div className="directory-list-toolbar">
        <div className="directory-list-toolbar-top">
          <div className="directory-list-toolbar-device">
            <span className="directory-list-toolbar-device-model">
              {device ? device.properties.model.replace(/_/g, " ") : null}
            </span>
            <span className="directory-list-toolbar-device-serial">
              {this.props.device}
            </span>
          </div>
          <div className="directory-list-toolbar-download">
            <button
              onClick={this.props.onDownloadFolderClick}
              className="directory-list-toolbar-download-button">
              <i className="fas fa-download directory-list-toolbar-download-icon" />
              Download this folder
            </button>
            <div className="directory-list-toolbar-path-destination">
              Files will be downloaded to{" "}
              <a
                className="directory-list-toolbar-path-link"
                href="#changepath"
                onClick={DeviceToolbar.handleDownloadPathClick}>
                {this.props.localPath}
              </a>
            </div>
          </div>
        </div>
        <div className="directory-list-toolbar-path">{renderedPath}</div>
      </div>
    );
  }
}

DeviceToolbar.defaultProps = {
  device: null
};

DeviceToolbar.propTypes = {
  device: DroidPropTypes.device,
  devicePath: DroidPropTypes.devicePath.isRequired,
  devices: DroidPropTypes.devices.isRequired,
  localPath: PropTypes.string,
  onDownloadFolderClick: PropTypes.func.isRequired,
  onBreadCrumbPathClick: PropTypes.func.isRequired
};

function mapStateToProps(state) {
  return {
    device: state.device,
    devicePath: state.devicePath,
    devices: state.devices,
    localPath: state.localPath
  };
}

function mapDispatchToProps(dispatch) {
  return {
    onDownloadFolderClick: () => dispatch(downloadFolder()),
    onBreadCrumbPathClick: path => {
      dispatch(setDevicePath(path));
    }
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(DeviceToolbar);
