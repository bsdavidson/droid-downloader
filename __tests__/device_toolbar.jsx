import React from "react";
import {shallow} from "enzyme";
import renderer from "react-test-renderer";

import {DeviceToolbar} from "../src/device_toolbar";
import {device, devicePath, devices} from "../__fixtures__/store";

test("<DeviceToolbar />", () => {
  const wrapper = renderer
    .create(
      <DeviceToolbar
        device={device}
        devicePath={devicePath}
        devices={devices}
        onDownloadFolderClick={() => {}}
      />
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});

test("<DeviceToolbar /> should render nothing without a device", () => {
  const wrapper = renderer
    .create(
      <DeviceToolbar
        device={""}
        devicePath={devicePath}
        devices={devices}
        onDownloadFolderClick={() => {}}
      />
    )
    .toJSON();
  expect(wrapper).toMatchSnapshot();
});

test("<DeviceToolbar /> should call onDownloadFolderClick", () => {
  const onDownloadFolderClick = jest.fn();

  const wrapper = shallow(
    <DeviceToolbar
      device={device}
      devicePath={devicePath}
      devices={devices}
      onDownloadFolderClick={onDownloadFolderClick}
    />
  );
  expect(onDownloadFolderClick.mock.calls.length).toBe(0);
  wrapper.find(".directory-list-toolbar-download-button").simulate("click");
  expect(onDownloadFolderClick.mock.calls.length).toBe(1);
});
