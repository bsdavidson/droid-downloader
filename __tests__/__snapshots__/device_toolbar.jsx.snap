// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeviceToolbar /> 1`] = `
<div>
  <div
    className="directory-list-toolbar"
  >
    <div
      className="directory-list-toolbar-device"
    >
      <div>
        <span
          className="directory-list-toolbar-device-model"
        >
          SM G999V
        </span>
        <span
          className="directory-list-toolbar-device-serial"
        >
          ffffbeefface
        </span>
      </div>
      <div
        className="directory-list-toolbar-path"
      >
        Folder listing for /
        storage/emulated/0/DCIM
      </div>
    </div>
    <div
      className="directory-list-toolbar-download"
    >
      <button
        className="directory-list-toolbar-download-button"
        onClick={[Function]}
      >
        <i
          className="fas fa-download directory-list-toolbar-download-icon"
        />
        Download this folder
      </button>
      <div
        className="directory-list-toolbar-path-destination"
      >
        Files will be downloaded to 
        /tmp/downloads
      </div>
    </div>
  </div>
</div>
`;

exports[`<DeviceToolbar /> should render nothing without a device 1`] = `null`;
